File: php/features.php
<?php
declare(strict_types=1);

// MetropolitanoDeLisboa SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class MetropolitanoDeLisboaFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new MetropolitanoDeLisboaBaseFeature();
            case "ratelimit":
                return new MetropolitanoDeLisboaRatelimitFeature();
            case "retry":
                return new MetropolitanoDeLisboaRetryFeature();
            case "test":
                return new MetropolitanoDeLisboaTestFeature();
            case "timeout":
                return new MetropolitanoDeLisboaTimeoutFeature();
            default:
                return new MetropolitanoDeLisboaBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/metropolitanodelisboa_sdk/features.py
# MetropolitanoDeLisboa SDK feature factory

from metropolitanodelisboa_sdk.feature.base_feature import MetropolitanoDeLisboaBaseFeature
from metropolitanodelisboa_sdk.feature.ratelimit_feature import MetropolitanoDeLisboaRatelimitFeature
from metropolitanodelisboa_sdk.feature.retry_feature import MetropolitanoDeLisboaRetryFeature
from metropolitanodelisboa_sdk.feature.test_feature import MetropolitanoDeLisboaTestFeature
from metropolitanodelisboa_sdk.feature.timeout_feature import MetropolitanoDeLisboaTimeoutFeature


_FEATURES = {
    "base": lambda: MetropolitanoDeLisboaBaseFeature(),
    "ratelimit": lambda: MetropolitanoDeLisboaRatelimitFeature(),
    "retry": lambda: MetropolitanoDeLisboaRetryFeature(),
    "test": lambda: MetropolitanoDeLisboaTestFeature(),
    "timeout": lambda: MetropolitanoDeLisboaTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
